### PR TITLE
New DAC-Related Transaction Types and API Requests

### PR DESCRIPTION
**Pull request by aeddaqqa**
## Extend CaminoJS with New DAC-Related Transaction Types and API Requests

This pull request introduces enhancements to the CaminoJS library by adding new transaction types and API requests related to DAC. The new transaction types included are AddProposalTx and AddVoteTx, aimed at streamlining proposal submission and voting processes within DACs.

#### Features:
1. **AddProposalTx**
   - Introduced a new transaction type, `AddProposalTx`, which allows the addition of proposals.
  
2. **AddVoteTx**
   - Added a new transaction type, `AddVoteTx`, to facilitate voting on proposals.
  
3. **Proposal Types**
   - Added support for two new proposal types: 
     - `NewMember`: For proposing the addition of new members.
     - `ExcludeMember`: For proposing the exclusion of existing members.
  
4. **ExcludeMemberProposal**
   - Implemented support for the new `ExcludeMemberProposal` type.
  
5. **AdminProposal**
   - Introduced a new proposal type, `AdminProposal`
  
6. **Address State**
   - Added a new address state specific to Camino-only proposers.

#### Fixes:
1. **Multisig Transaction Issues**
   - Resolved issues related to creating multisig transactions for both `AddProposal` and `AddVote` transaction types.

**Changes:**
- **feat(addproposal):** Added `AddProposalTx`.
- **feat(addvote):** Added `AddVoteTx`.
- **feat(proposal):** Added `NewMember` and `ExcludeMember` proposal types.
- **feat(exludememberproposal):** Added support for `ExcludeMemberProposal`.
- **feat(adminproposal):** Added `AdminProposal` type.
- **feat(adminProposal):** Added new proposal type `adminProposal`.
- **feat(addressState):** Added new address state for Camino-only proposer.
- **fix(addProposal, addVote):** Resolved issues creating multisig transactions.

### CODE REVIEW

File: src/apis/platformvm/addvotetx.ts [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+    )
+    this.voterAuth.deserialize(fields["voterAuth"], encoding)
+    this.votePayload = this.votePayload.deserialize(fields, encoding)
+  }
+
+  protected upgradeVersionID = new UpgradeVersionID()
+  protected proposalID = Buffer.alloc(32)
+  protected votePayload: VoteWrapper
+  protected voterAddress = Buffer.alloc(20)
+  protected voterAuth: SubnetAuth
```
Comment (knikos) [inline src/apis/platformvm/addvotetx.ts]: can you move these properties further up right after typeID  (for easier readability)

File: src/apis/platformvm/api.ts [hunk 2]
```diff
@@ -3145,4 +3147,154 @@ export class PlatformVMAPI extends JRPCAPI {
...
+
+    return builtUnsignedTx
+  }
+
+  /**
+   * Build an unsigned [[AddVoteTx]].
+   *
+   * @param utxoset A set of UTXOs that the transaction is built on
+   * @param fromAddresses The addresses being used to send the funds from the UTXOs {@link https://github.com/feross/buffer|Buffer}
+   * @param changeAddresses The addresses that can spend the change remaining from the spent UTXOs.
```
Comment (knikos) [inline src/apis/platformvm/api.ts:3231]: missing  a docu line for parameter `proposalID` here

File: src/apis/platformvm/builder.ts [hunk 3]
```diff
@@ -1786,4 +1789,170 @@ export class Builder {
...
+    return new UnsignedTx(baseTx)
+  }
+
+  /**
+   * Build an unsigned [[AddVoteTx]].
+   *
+   * @param networkID Networkid, [[DefaultNetworkID]]
+   * @param blockchainID Blockchainid, default undefined
+   * @param fromSigner The addresses being used to send and verify the funds from the UTXOs {@link https://github.com/feross/buffer|Buffer}
+   * @param changeAddresses The addresses that can spend the change remaining from the spent UTXOs.
```
Comment (knikos) [inline src/apis/platformvm/builder.ts:1881]: missing a docu line for parameter proposalID here

File: src/apis/platformvm/addproposaltx/addmemberproposal.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    endTime.writeUInt32BE(end, 4)
+    super(startTime, endTime)
+
+    if (typeof applicantAddress === "string") {
+      this.applicantAddress = bintools.stringToAddress(applicantAddress)
+    } else {
+      this.applicantAddress = applicantAddress
+    }
+
+    // // Initialize options
```
Comment (knikos) [inline src/apis/platformvm/addproposaltx/addmemberproposal.ts]: i suppose these are just leftovers. so please delete.

File: src/apis/platformvm/addproposaltx/index.ts [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+  }
+
+  /**
+   * Class representing an unsigned RegisterNode transaction.
+   *
+   * @param networkID Optional networkID, [[DefaultNetworkID]]
+   * @param blockchainID Optional blockchainID, default Buffer.alloc(32, 16)
+   * @param outs Optional array of the [[TransferableOutput]]s
+   * @param ins Optional array of the [[TransferableInput]]s
+   * @param memo Optional {@link https://github.com/feross/buffer|Buffer} for the memo field
```
Comment (knikos) [inline src/apis/platformvm/addproposaltx/index.ts:311]: add docu for proposalDescription

File: tests/apis/platformvm/addproposaltx.test.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    const addProposalTxTypeName: string = addProposalTx.getTypeName()
+    expect(addProposalTxTypeName).toBe("AddProposalTx")
+  })
+
+  test("getTypeID", async (): Promise<void> => {
+    const addProposalTxTypeID: number = addProposalTx.getTypeID()
+    expect(addProposalTxTypeID).toBe(PlatformVMConstants.ADDPROPOSALTX)
+    console.log("addProposalTx decoded: ", addProposalTx)
+  })
+
```
Comment (knikos) [inline tests/apis/platformvm/addproposaltx.test.ts:169]: pleases add tests for `getProposalPayload` , `getProposalDescription` and `getProposerAuth`

**Review by knikos — DISMISSED** [review]
see comments

File: src/apis/platformvm/addproposaltx/excludememberproposal.ts
Comment (VjeraTurk) [inline src/apis/platformvm/addproposaltx/excludememberproposal.ts:1]: Comment present

File: tests/apis/platformvm/addproposaltx.test.ts (hunk 6 above)
Comment (VjeraTurk) [reply tests/apis/platformvm/addproposaltx.test.ts:169]: getProposerAuth test missing? @aeddaqqa 
Maybe not necessary?

**Review by VjeraTurk — APPROVED** [review]

**Review by VjeraTurk — APPROVED** [review]

**Pull request merged**